### PR TITLE
Use a single index Results when querying across blocks

### LINKED ISSUE #1469: Use a single index Results when querying across blocks
Right now, when we query across many index blocks we allocate a results map for each one, run all the queries in parallel for each block filling up the results map, and then we merge all the results maps together at the end. This makes parallelization easy and straightforward, but since many workloads contain many of the same documents in each block, it causes memory utilization to scale linearly with the number of index blocks. This is problematic for queries that return a large number of documents.

Screen Shot 2019-03-17 at 10 21 11 AM

To address this issue, we will use a single index Results which we will push down into all the block queries. That way, each document will only get allocated in memory once regardless of the number of blocks that are being queried. This will significantly improve memory utilization but will make the results map very heavily contended. To address that, instead of acquiring a lock each time to add a single document, we'll use pooled slices to temporarily gather a configurable number of documents together within the block, then acquire the lock once and add them all together as a batch.

This should give us significantly reduced memory utilization without causing too much contention on the lock.

### PR DESCRIPTION
**Pull request by robskillington**
Fixes #1469

### CODE REVIEW

File: src/dbnode/storage/index.go [hunk 1]
```diff
@@ -1064,8 +1017,6 @@ func (i *nsIndex) Query(
 	}
 
 	results.Lock()
-	// Signal not to add any further results since we've returned already.
-	results.returned = true
```
Comment (richardartoul) [inline src/dbnode/storage/index.go]: Are we losing the ability to do this signaling? Seems like this might make the impact of timed out queries even worse since they'll keep updating this map...Actually isn't it broken because they'll keep trying to add results to a map that may have been returned to the pool?

File: src/dbnode/storage/index/block_test.go [hunk 2]
```diff
@@ -510,13 +510,14 @@ func TestBlockMockQueryExecutorExecLimit(t *testing.T) {
...
 	require.False(t, exhaustive)
 
-	rMap := results.Map()
-	require.Equal(t, 1, rMap.Len())
-	t1, ok := rMap.Get(ident.StringID(string(testDoc1().ID)))
-	require.True(t, ok)
-	require.True(t, ident.NewTagIterMatcher(
-		ident.MustNewTagStringsIterator("bar", "baz")).Matches(
-		ident.NewTagsIterator(t1)))
+	results.WithMap(func(rMap *ResultsMap) {
```
Comment (arnikola) [inline src/dbnode/storage/index/block_test.go]: nit: this is used in multiple places, maybe make it into a test function?

File: src/dbnode/storage/index/block.go [hunk 3]
```diff
@@ -772,23 +774,50 @@ func (b *block) Query(
...
+
+		_, size, err = results.AddDocumentsBatch(batch, batchOpts)
+		if err != nil {
+			return false, err
+		}
+
+		// Reset batch
+		var emptyDoc doc.Document
+		for i := range batch {
+			batch[i] = emptyDoc
```
Comment (arnikola) [inline src/dbnode/storage/index/block.go]: Worth doing some sort of Memset optimization here instead?

File: src/dbnode/storage/index/block.go [hunk 4]
```diff
@@ -772,23 +774,50 @@ func (b *block) Query(
...
+	}
+	batchOpts := AddDocumentsBatchResultsOptions{
+		AllowPartialUpdates: true,
+		LimitSize:           opts.Limit,
+	}
 	iterCloser := safeCloser{closable: iter}
 	execCloser := safeCloser{closable: exec}
 
 	defer func() {
+		b.docsPool.Put(batch)
```
Comment (arnikola) [inline src/dbnode/storage/index/block.go]: Do we need to reset the batch here?

File: src/dbnode/storage/index/results.go [hunk 5]
```diff
@@ -22,17 +22,20 @@ package index
...
 )
 
 var (
-	errUnableToAddResultMissingID = errors.New("no id for result")
+	errUnableToAddResultMissingID          = errors.New("no id for result")
+	errResultAlreadyExistsNoPartialUpdates = errors.New("id already exists for result and partial updates not allowed")
 )
 
 type results struct {
+	sync.RWMutex
```
[the remark this answers is not in the document]
Comment (prateek) [reply src/dbnode/storage/index/results.go]: depends on access pattern mainly: A RWMutex is useful when you have potential multiple readers that could share underlying data without clobbering each other; if all accessors always need r/w access - then a mutex is fine.

[the remark this answers is not in the document]
Comment (arnikola) [reply src/dbnode/storage/index/results.go]: Sorry, meant something like

```golang
type results struct {
 sync.RWMutex
}
```

v.s.

```golang
type results struct {
 mu sync.RWMutex
}
```

File: src/dbnode/storage/index/types.go [hunk 6]
```diff
@@ -79,13 +79,17 @@ type QueryResults struct {
...
 	// Namespace returns the namespace associated with the result.
 	Namespace() ident.ID
 
-	// Map returns a map from seriesID -> seriesTags, comprising index results.
-	Map() *ResultsMap
+	// WithMap executes a function that has access to a results map
+	// from seriesID -> seriesTags, comprising index results.
+	// A function is required to ensure that the results is used
+	// while a read lock for the results is held.
+	WithMap(fn func(results *ResultsMap))
```
Comment (prateek) [inline src/dbnode/storage/index/types.go]: instead of this, how would you feel about adding a `Seal()` and changing `Map()` to return `(*ResultsMap, error)`. Would allow callers to avoid the extra lambda.

File: src/dbnode/storage/index/results.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: I prefer the prior, unless it's an exported struct in which case the latter to reduce scope of calling those methods.

File: src/dbnode/storage/index/results.go [hunk 7]
```diff
@@ -55,6 +58,15 @@ func NewResults(opts Options) Results {
 
 func (r *results) AddDocument(
 	d doc.Document,
+) (added bool, size int, err error) {
+	r.Lock()
+	added, size, err = r.addDocumentWithLock(d)
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: This is to keep the existing stuff here.  Usually we prefer to return structs actually instead of named returns.

File: src/dbnode/storage/index/results.go [hunk 8]
```diff
@@ -82,9 +94,55 @@ func (r *results) AddDocument(
...
+		}
+		if opts.LimitSize > 0 && size >= opts.LimitSize {
+			// Early return if limit enforced and we hit our limit
+			break
+		}
+	}
+	return numPartialUpdates, size, nil
+}
+
 func (r *results) AddIDAndTags(
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Yeah true, will do - need to remove the tests.

File: src/dbnode/storage/index/results.go [hunk 9]
```diff
@@ -121,18 +179,29 @@ func (r *results) cloneTagsFromFields(fields doc.Fields) ident.Tags {
...
-	return r.nsID
+	r.RLock()
+	v := r.nsID
+	r.RUnlock()
+	return v
 }
 
-func (r *results) Map() *ResultsMap {
-	return r.resultsMap
+func (r *results) WithMap(fn func(results *ResultsMap)) {
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure thing.

File: src/dbnode/storage/index/results.go [hunk 10]
```diff
@@ -121,18 +179,29 @@ func (r *results) cloneTagsFromFields(fields doc.Fields) ident.Tags {
...
+	r.RLock()
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Yeah I'll rename it, maybe add a with lock version too.

File: src/dbnode/storage/index/results.go [hunk 11]
```diff
@@ -121,18 +179,29 @@ func (r *results) cloneTagsFromFields(fields doc.Fields) ident.Tags {
...
-	return r.resultsMap.Len()
+	r.RLock()
+	v := r.resultsMap.Len()
+	r.RUnlock()
+	return v
 }
 
 func (r *results) Reset(nsID ident.ID) {
+	r.Lock()
+	defer r.Unlock()
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure, probably don't need it here.

File: src/dbnode/storage/index/results.go [hunk 12]
```diff
@@ -157,7 +226,11 @@ func (r *results) Reset(nsID ident.ID) {
 }
 
 func (r *results) Finalize() {
-	if r.noFinalize {
+	r.RLock()
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: True, but I figured I was locking everything and it is kind of strange to lock only some state rather than all.

File: src/dbnode/storage/index/results.go [hunk 13]
```diff
@@ -157,7 +226,11 @@ func (r *results) Reset(nsID ident.ID) {
...
+	noFinalize := r.noFinalize
+	r.RUnlock()
+
+	if noFinalize {
 		return
 	}
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: When putting back in the pool?  Sure I'm not opposed to that, but technically just before you add it to the pool you're basically declaring you're done with the object.

File: src/dbnode/storage/index/results.go [hunk 14]
```diff
@@ -170,6 +243,9 @@ func (r *results) Finalize() {
 }
 
 func (r *results) NoFinalize() {
+	r.Lock()
+	defer r.Unlock()
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure, probably don't need it here.

File: src/dbnode/storage/index/types.go (hunk 6 above)
Comment (robskillington) [reply src/dbnode/storage/index/types.go]: Yeah, I think this is actually the approach I'll take also to address the pooling issues that Richie was concerned about (i.e. not knowing when it's finalized and potentially adding values to it after it is finalized)

Comment (richardartoul) [reply src/dbnode/storage/index/types.go]: @robskillington I think if you have an early time out / cancel, you have to give up and not return it to the pool cause even if you seal it there is no way to know when its safe to unseal it unless you ref count it

File: src/dbnode/storage/index/results.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (richardartoul) [reply src/dbnode/storage/index/results.go]: yeah I dont have strong feelings either way, just took me a second to think about why you might be doing it

File: src/dbnode/storage/index/results.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: On second thoughts, can't lock during entire call to Finalize as Reset itself acquires a write lock, that's why we release the lock earlier in the method.

File: src/dbnode/storage/index/block.go (hunk 4 above)
Comment (robskillington) [reply src/dbnode/storage/index/block.go]: No thankfully the Put method on the array will zero out each elem and also resize it to zero.

File: src/dbnode/storage/index/block.go (hunk 3 above)
Comment (robskillington) [reply src/dbnode/storage/index/block.go]: This is the memset optimization thankfully, just using a single zero var value and then setting everything with `range i := ... ` to that value.

File: src/dbnode/storage/index/results.go [hunk 15]
```diff
@@ -170,6 +243,9 @@ func (r *results) Finalize() {
...
+
 	// Ensure neither the results object itself, or any of its underlying
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure thing.

File: src/dbnode/storage/index.go [hunk 16]
```diff
@@ -895,24 +895,16 @@ func (i *nsIndex) Query(
 			exhaustive bool
 			returned   bool
 		}{
-			merged:     nil,
+			merged:     i.resultsPool.Get(),
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: Sure thing, renamed.

File: src/dbnode/storage/index.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/storage/index.go]: So I now use a lifetime to protect against writing to the results once we return from the Query call to the index, this prevents writing to results during cancellation or any other early return code path.

File: src/dbnode/storage/index/types.go (hunk 6 above)
Comment (robskillington) [reply src/dbnode/storage/index/types.go]: So I now use a lifetime to protect against writing to the results once we return from the Query call to the index, this prevents writing to results during cancellation or any other early return code path.

File: src/dbnode/server/server.go [hunk 17]
```diff
@@ -1223,7 +1222,11 @@ func withEncodingAndPoolingOptions(
 		SetCheckedBytesPool(bytesPool).
 		SetResultsPool(resultsPool)
 
-	resultsPool.Init(func() index.Results { return index.NewResults(indexOpts) })
+	resultsPool.Init(func() index.Results {
+		// NB(r): Need to initialize after setting the index opts so
+		// it seems the same reference of the options as is set
```
Comment (richardartoul) [inline src/dbnode/server/server.go]: This sentence reads like word salad lol. I think you were trying to say: "so it sees the same reference to the indexOptions"

File: src/x/resource/lifetime.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+// NewCancellableLifetime returns a new cancellable resource lifetime.
+func NewCancellableLifetime() *CancellableLifetime {
+	return &CancellableLifetime{}
+}
+
+// TryCheckout will try to checkout the resource, if the lifetime
+// is already cancelled this will return false, otherwise it will return
+// true and guarantee the lifetime is not cancelled until the checkout
+// is returned.
+func (l *CancellableLifetime) TryCheckout() bool {
```
Comment (richardartoul) [inline src/x/resource/lifetime.go]: Can you add to the comment: If this returns true you MUST call `ReleaseCheckout` later, but if it returns false calling `ReleaseCheckout` will panic (pretty sure unlocking an unlocked lock panics

File: src/dbnode/storage/index/compaction/compactor.go [hunk 19]
```diff
@@ -250,6 +254,36 @@ func (c *Compactor) compactFromBuilderWithLock(
...
 
+	if !c.opts.MmapDocsData {
+		// If retaining references to the original docs, simply take ownership
+		// of the documents and then reference them directly from the FST segment
+		// rather than encoding them and mmap'ing the encoded documents.
+		allDocsCopy := make([]doc.Document, len(allDocs))
+		copy(allDocsCopy, allDocs)
+		fstData.DocsReader = docs.NewSliceReader(0, allDocsCopy)
+	} else {
+		// Otherwise encode and reference the encoded bytes as mmap'd bytes.
```
Comment (arnikola) [inline src/dbnode/storage/index/compaction/compactor.go]: I'm not really familiar with this path, but just verifying it's ok to have DocsReader be nil here?

File: src/dbnode/storage/index.go [hunk 20]
```diff
@@ -921,75 +919,33 @@ func (i *nsIndex) Query(
...
-				id, tags := entry.Key(), entry.Value()
-				_, size, err = results.merged.AddIDAndTags(id, tags)
-				if err != nil {
-					results.multiErr = results.multiErr.Add(err)
-					return
-				}
-			}
-		}
-
 		// If block had more data but we stopped early, need to notify caller.
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: Sounds good, will do.

File: src/dbnode/storage/index.go [hunk 21]
```diff
@@ -921,75 +919,33 @@ func (i *nsIndex) Query(
...
+		alreadyNotExhaustive := opts.LimitExceeded(size)
 		if alreadyNotExhaustive {
-			results.exhaustive = false
+			state.Lock()
+			state.exhaustive = false
+			state.Unlock()
 		}
-		results.Unlock()
 
 		if alreadyNotExhaustive {
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: Sure thing, this is a side effect of the refactor actually, good catch.

File: src/dbnode/storage/index.go [hunk 22]
```diff
@@ -921,75 +919,33 @@ func (i *nsIndex) Query(
...
 		}
-		results.exhaustive = false
+		state.exhaustive = false
 	}
 
 	for _, block := range blocks {
 		// Capture block for async query execution below.
 		block := block
 
 		// Terminate early if we know we don't need any more results.
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: Thank you for writing the comment, will do - hah 👍

File: src/dbnode/storage/index.go [hunk 23]
```diff
@@ -1063,29 +1019,20 @@ func (i *nsIndex) Query(
 		}
 	}
 
-	results.Lock()
-	// Signal not to add any further results since we've returned already.
-	results.returned = true
+	state.Lock()
 	// Take reference to vars to return while locked, need to allow defer
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: I'll remove that statement, there used to be other stuff going on.

File: src/dbnode/storage/index/block.go [hunk 24]
```diff
@@ -806,10 +827,39 @@ func (b *block) Query(
...
+func (b *block) addQueryResults(
+	cancellable *resource.CancellableLifetime,
+	results Results,
+	batch []doc.Document,
+) ([]doc.Document, int, error) {
+	// Checkout the lifetime of the query before adding results
+	queryValid := cancellable.TryCheckout()
+	if !queryValid {
+		// Query not valid any longer, do not add results and return early
+		return batch, 0, errCancelledQuery
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/block.go]: Yeah, we always lock when we access multierr.

File: src/dbnode/storage/index/results.go [hunk 25]
```diff
@@ -44,19 +49,77 @@ type results struct {
...
+		pool:       indexOpts.ResultsPool(),
+	}
+}
+
+func (r *results) Reset(nsID ident.ID, opts ResultsOptions) {
+	r.Lock()
+
+	r.opts = opts
+
+	// finalize existing held nsID
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure thing, this is old code but I'll update it.

[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: PS we should really create a lint rule so you don't have to manually ask this btw (not urgent, but something we should think about)

File: src/dbnode/storage/index/results.go [hunk 26]
```diff
@@ -44,19 +49,77 @@ type results struct {
...
+	}
+	return nil
+}
+
+func (r *results) addDocumentWithLock(
 	d doc.Document,
-) (added bool, size int, err error) {
-	added = false
+) (bool, int, error) {
+	added := false
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: Sure

File: src/dbnode/storage/index/types.go [hunk 27]
```diff
@@ -79,45 +80,46 @@ type QueryResults struct {
...
+	// index results.
 	Map() *ResultsMap
 
-	// Reset resets the Results object to initial state.
-	Reset(nsID ident.ID)
+	// AddDocuments adds the batch of documents to the results set, it will
+	// take a copy of the bytes backing the documents so the original can be
+	// modified after this function returns without affecting the results map.
+	// If documents with duplicate IDs are added, they are simply ignored and
+	// the first document added with an ID is returned.
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/types.go]: Sure thing.

File: src/x/resource/lifetime.go (hunk 18 above)
Comment (robskillington) [reply src/x/resource/lifetime.go]: Yeah sure thing.

Comment (robskillington) [reply src/x/resource/lifetime.go]: Added comments.

File: src/dbnode/storage/index.go [hunk 28]
```diff
@@ -921,75 +924,33 @@ func (i *nsIndex) Query(
...
 
 		// Terminate early if we know we don't need any more results.
-		results.Lock()
-		mergedSize := 0
-		if results.merged != nil {
-			mergedSize = results.merged.Size()
-		}
-		alreadyNotExhaustive := opts.Limit > 0 && mergedSize >= opts.Limit
+		size := results.Size()
+		alreadyNotExhaustive := opts.LimitExceeded(size)
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index.go]: Good call.

File: src/dbnode/storage/index/block.go [hunk 29]
```diff
@@ -156,48 +160,58 @@ type blockShardRangesSegments struct {
...
 
 	backgroundCompactor := compaction.NewCompactor(docsPool,
 		documentArrayPoolCapacity,
-		opts.SegmentBuilderOptions(),
-		opts.FSTSegmentOptions(),
-		compaction.CompactorOptions{})
+		indexOpts.SegmentBuilderOptions(),
+		indexOpts.FSTSegmentOptions(),
+		compaction.CompactorOptions{
+			MmapDocsData: opts.ForegroundCompactorMmapDocsData,
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/block.go]: Hah, good call - TY for this catch.

File: src/dbnode/storage/index/types.go [hunk 30]
```diff
@@ -79,45 +80,46 @@ type QueryResults struct {
 	Exhaustive bool
 }
 
-// Results is a collection of results for a query.
+// Results is a collection of results for a query, it is synchronized
+// when access to the results set is used as documented by the methods.
+// It cannot be written to after it is sealed, until it's reopened by
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/types.go]: True, thanks will update.

File: src/dbnode/storage/index/results.go [hunk 31]
```diff
@@ -121,47 +157,37 @@ func (r *results) cloneTagsFromFields(fields doc.Fields) ident.Tags {
...
+	v := r.nsID
+	r.RUnlock()
+	return v
 }
 
 func (r *results) Map() *ResultsMap {
-	return r.resultsMap
+	r.RLock()
+	v := r.resultsMap
+	r.RUnlock()
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results.go]: I added a comment to the call to Map about this being unsafe.  We can't really take a copy since it will cause a ton of allocations.

File: src/dbnode/storage/index/results_test.go [hunk 32]
```diff
@@ -37,38 +38,46 @@ var (
 )
 
 func init() {
-	testOpts = NewOptions()
+	// Use size of 1 to test edge cases by default
+	testOpts = optionsWithDocsArrayPool(NewOptions(), 1, 1)
+}
+
+func optionsWithDocsArrayPool(opts Options, size, capacity int) Options {
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/index/results_test.go]: We call this method elsewhere.

File: src/dbnode/storage/index/compaction/compactor.go (hunk 19 above)
Comment (robskillington) [reply src/dbnode/storage/index/compaction/compactor.go]: Yup, if we supply the docs data and docs index data then its safe to have DocsReader be nil.
